### ISSUE TITLE
rOTA DE FUGA REVISAO 2

### ISSUE THREAD

**Issue by guilhermegandra** [guilhermegandra/Doutorado#2]
# Rotas de Fuga REV 02

!apt-get install -y libspatialindex-dev > /dev/null 2>&1
!pip install -q geopandas shapely fiona rtree pyproj tqdm networkx scipy scikit-learn openpyxl pandas

import os
import zipfile
import math
import time
import pandas as pd

import geopandas as gpd
import networkx as nx
import numpy as np
from shapely.geometry import Point, LineString
from shapely.ops import snap, split, unary_union, nearest_points
from tqdm import tqdm
from scipy.spatial import cKDTree
from sklearn.cluster import DBSCAN
from google.colab import files

# Funções de controle de tempo
def start_stage(nome):
    print(f"\nETAPA: {nome}")
    return time.perf_counter()

def end_stage(t0):
    dt = time.perf_counter() - t0
    print(f"Tempo: {dt:.1f} s ({dt/60:.2f} min)")

# Upload e extração de arquivos ZIP contendo SHP
def upload_and_extract(label):
    print(f"\nEnvie o ZIP contendo o shapefile de: {label}")
    uploaded = files.upload()
    zip_name = list(uploaded.keys())[0]

    extract_dir = f"input_{label}"
    os.makedirs(extract_dir, exist_ok=True)

    with zipfile.ZipFile(zip_name, "r") as z:
        z.extractall(extract_dir)

    shp = [f for f in os.listdir(extract_dir) if f.lower().endswith(".shp")]

    if not shp:
        raise ValueError(f"O ZIP enviado para {label} não contém arquivo .shp")

    return os.path.join(extract_dir, shp[0])

# Upload dos dados
t0 = start_stage("Upload dos shapefiles ZIP")
vias_path   = upload_and_extract("VIAS")
saidas_path = upload_and_extract("EDIFICACOES")
mancha_path = upload_and_extract("MANCHA_INUNDACAO")
end_stage(t0)

# Leitura dos shapefiles
t0 = start_stage("Leitura dos dados")
vias   = gpd.read_file(vias_path)
saidas = gpd.read_file(saidas_path)
mancha = gpd.read_file(mancha_path)
end_stage(t0)

# Garantir CRS
def garantir_crs(gdf, nome):
    if gdf.crs is None:
        print(f"O arquivo '{nome}' não possui CRS.")
        epsg = int(input(f"Informe o EPSG original de {nome}: "))
        gdf = gdf.set_crs(epsg)
    return gdf

# Ajuste CRS
t0 = start_stage("Reprojeção para EPSG:31982 (SIRGAS 2000 / UTM 22S)")
vias   = garantir_crs(vias,   "vias").to_crs(31982)
saidas = garantir_crs(saidas, "edificacoes").to_crs(31982)
mancha = garantir_crs(mancha, "mancha").to_crs(31982)
end_stage(t0)

# Velocidade
t0 = start_stage("Parâmetros de deslocamento")
vel_kmh = float(input("Velocidade média (km/h): "))
vel_ms = vel_kmh * 1000 / 3600
end_stage(t0)

# Limpeza da rede viária
t0 = start_stage("Limpeza e normalização da rede viária")

vias = vias[~vias.geometry.is_empty & ~vias.geometry.isna()].copy()
vias["geometry"] = vias.geometry.apply(lambda g: g if g.is_valid else g.buffer(0))
vias = vias[vias.geometry.type.isin(["LineString", "MultiLineString"])].copy()

linhas = []
for geom in tqdm(vias.geometry, desc="Explodindo geometrias", ncols=80):
    if geom.geom_type == "LineString":
        linhas.append(geom)
    else:
        linhas.extend(list(geom.geoms))

todos_pontos = [Point(c[:2]) for l in linhas for c in l.coords]
uniao_pontos = gpd.GeoSeries(todos_pontos, crs=vias.crs).unary_union

snapped = []
for l in tqdm(linhas, desc="Aplicando SNAP", ncols=80):
    snapped.append(snap(l, uniao_pontos, tolerance=0.5))

vias_split = []
for i in tqdm(range(len(snapped)), desc="Split", ncols=80):
    sline = snapped[i]
    for j in range(len(snapped)):
        if i != j and sline.intersects(snapped[j]):
            try:
                sline = split(sline, snapped[j])
            except:
                pass
    if sline.geom_type in ["MultiLineString", "GeometryCollection"]:
        vias_split.extend([seg for seg in sline.geoms if seg.length > 0])
    else:
        if sline.length > 0:
            vias_split.append(sline)

end_stage(t0)

# Construção do grafo
t0 = start_stage("Construção do grafo viário")

G = nx.Graph()
for line in tqdm(vias_split, desc="Criando arestas", ncols=80):
    coords = [tuple(c[:2]) for c in line.coords]
    dist_m = line.length
    tempo_min = dist_m / vel_ms / 60
    tempo_min = max(tempo_min, 0.0001)
    seg_tempo = tempo_min / (len(coords) - 1)

    for i in range(len(coords) - 1):
        G.add_edge(coords[i], coords[i + 1], weight=seg_tempo)

node_coords = np.array(list(G.nodes))
tree = cKDTree(node_coords)

def nearest_node(pt):
    _, idx = tree.query([pt.x, pt.y])
    return tuple(node_coords[idx])

end_stage(t0)

# Seleção dos imóveis afetados
t0 = start_stage("Identificação de edificações afetadas")
mancha_union = mancha.geometry.unary_union
saidas_afetadas = gpd.sjoin(saidas, mancha, how="inner", predicate="intersects").reset_index(drop=True)
end_stage(t0)

# Função: procurar nó fora da mancha
def nearest_node_outside_flood(point, tree, node_coords, mancha_union):
    for k in [1,3,5,10,20,50,100,200]:
        _, idxs = tree.query([point.x, point.y], k=k)
        idxs = [idxs] if k == 1 else idxs
        for idx in idxs:
            cand = Point(node_coords[idx])
            if not cand.within(mancha_union):
                return tuple(node_coords[idx])
    raise ValueError("Nenhum nó viário fora da mancha encontrado.")

# Ponto global
t0 = start_stage("Cálculo do ponto global de encontro")

centro = saidas_afetadas.geometry.unary_union.centroid
if centro.within(mancha_union):
    pteo = nearest_points(centro, mancha_union.boundary)[1]
else:
    pteo = centro

dest_global = nearest_node_outside_flood(pteo, tree, node_coords, mancha_union)
ponto_global = Point(dest_global[0], dest_global[1])

end_stage(t0)

# Rotas iniciais
t0 = start_stage("Cálculo das rotas iniciais")

rotas_ini = []
tempos_ini = []
rota_ok_ini = []

for idx, row in tqdm(saidas_afetadas.iterrows(), total=len(saidas_afetadas), desc="Rotas"):
    try:
        start = nearest_node(row.geometry.centroid)
        tempo = nx.shortest_path_length(G, start, dest_global, weight="weight")
        tempo = math.ceil(tempo)
        path = nx.shortest_path(G, start, dest_global, weight="weight")
        rotas_ini.append(LineString(path))
        tempos_ini.append(tempo)
        rota_ok_ini.append(True)
    except:
        rotas_ini.append(None)
        tempos_ini.append(None)
        rota_ok_ini.append(False)

saidas_afetadas["tempo_min"] = tempos_ini
saidas_afetadas["rota_ok"] = rota_ok_ini

end_stage(t0)

# Tempo máximo
t0 = start_stage("Aplicação do tempo máximo permitido")
tempo_max = float(input("Tempo máximo permitido (min): "))
indices_excedidos = [i for i,t in enumerate(tempos_ini) if t is None or t > tempo_max]
end_stage(t0)

# DBSCAN 500 m
t0 = start_stage("Agrupamento DBSCAN para reduzir pontos de encontro")

pontos_teo = []
for i in indices_excedidos:
    g = saidas_afetadas.loc[i, "geometry"]
    c = g.centroid
    if c.within(mancha_union):
        c = nearest_points(c, mancha_union.boundary)[1]
    pontos_teo.append(c)

if pontos_teo:
    coords = np.array([[p.x, p.y] for p in pontos_teo])
    db = DBSCAN(eps=500, min_samples=1).fit(coords)
    labels = db.labels_
    clusters = sorted(set(labels))
else:
    labels = []
    clusters = []

end_stage(t0)

# Construção final das rotas considerando clusters
t0 = start_stage("Cálculo das rotas finais dos clusters")

rotas_final = rotas_ini.copy()
tempos_final = tempos_ini.copy()
id_ponto_final = [0 if tempos_ini[i] is not None and tempos_ini[i] <= tempo_max else None
                  for i in range(len(saidas_afetadas))]

pontos_encontro = [ponto_global]

for lab in clusters:
    cluster_indices = [indices_excedidos[k] for k,lb in enumerate(labels) if lb==lab]
    pts_cluster = [pontos_teo[k] for k,lb in enumerate(labels) if lb==lab]

    centro_cluster = unary_union(pts_cluster).centroid
    dest_cluster = nearest_node_outside_flood(centro_cluster, tree, node_coords, mancha_union)
    pcluster = Point(dest_cluster[0], dest_cluster[1])

    pontos_encontro.append(pcluster)
    pid = len(pontos_encontro)-1

    for i in cluster_indices:
        try:
            start = nearest_node(saidas_afetadas.loc[i, "geometry"].centroid)
            tempo = nx.shortest_path_length(G, start, dest_cluster, weight="weight")
            tempo = math.ceil(tempo)
            path = nx.shortest_path(G, start, dest_cluster, weight="weight")

            rotas_final[i] = LineString(path)
            tempos_final[i] = tempo
            id_ponto_final[i] = pid
        except:
            rotas_final[i] = None
            tempos_final[i] = None
            id_ponto_final[i] = None

saidas_afetadas["tempo_min"] = tempos_final
saidas_afetadas["id_ponto"] = id_ponto_final

end_stage(t0)

# Exportação
t0 = start_stage("Exportação dos resultados")

os.makedirs("saida_colab/saida_final", exist_ok=True)

# Rotas
gpd.GeoDataFrame(
    {"tempo_min": tempos_final, "id_ponto": id_ponto_final},
    geometry=rotas_final,
    crs=vias.crs
).to_file("saida_colab/saida_final/rotas.shp")

# Edificações
saidas_afetadas.to_file("saida_colab/saida_final/saidas_afetadas.shp")

# Pontos de encontro
gpd.GeoDataFrame(
    {"id": list(range(len(pontos_encontro)))},
    geometry=pontos_encontro,
    crs=vias.crs
).to_file("saida_colab/saida_final/pontos_encontro.shp")

# XLSX 
df_xlsx = pd.DataFrame({
    "id_imovel": saidas_afetadas.index,
    "id_ponto_encontro": id_ponto_final,
    "tempo_min": tempos_final,
    "rota_ok": [r is not None for r in rotas_final]
})
df_xlsx.to_excel("saida_colab/saida_final/imoveis_pontos.xlsx", index=False)

# ZIP final
zip_name = "resultados_rotas.zip"
with zipfile.ZipFile(zip_name, "w", zipfile.ZIP_DEFLATED) as z:
    for root, _, files_list in os.walk("saida_colab"):
        for file in files_list:
            full_path = os.path.join(root, file)
            arcname = os.path.relpath(full_path, "saida_colab")
            z.write(full_path, arcname)

files.download(zip_name)

end_stage(t0)

print("\nProcessamento finalizado.")
